src/cljam/io/bam_index/writer.clj

```diff
@@ -310,7 +228,127 @@
+            (write-bin wtr (:bin bin) (:chunks bin)))
+            ;; meta data
+          (write-meta-data wtr (:meta-data index))))
+        ;; linear index
+      (lsb/write-int wtr (count (:linear-index index)))
+      (doseq [l (:linear-index index)]
+        (lsb/write-long wtr l))))
+    ;; no coordinate alignments
+  (lsb/write-long wtr (:no-coordinate-alns indices)))
+
+;; ### Making index
+
+(defn make-index*
+  "Calculates index from the references and alignments, returning it as a map.
+  Returned index is still intermediate. It must be passed to finalize function
+  in the final stage."
+  [alns]
+  (loop [[^BAMPointerBlock aln & rest] alns
+         rid (.ref-id aln)
+         idx-status (init-index-status)
+         no-coordinate-alns 0
+         indices {}]
+    (if aln
+      (let [rid' (.ref-id aln)
+            new-ref? (not= rid' rid)
+            idx-status' (update-index-status
+                         (if new-ref? (init-index-status) idx-status) aln)
+            no-coordinate-alns' (if (zero? (.pos aln))
+                                  (inc no-coordinate-alns)
+                                  no-coordinate-alns)
+            indices' (if new-ref?
+                       (assoc indices rid idx-status)
+                       indices)]
+        (recur rest rid' idx-status' no-coordinate-alns' indices'))
+      (assoc indices rid idx-status
+             :no-coordinate-alns no-coordinate-alns))))
+
+(defn update-lastpointer-index
+  "Update last pointer of intermediate BAM index data structure."
+  [^long nrefs index ^long oldpt ^long newpt]
+  (let [update-pt-fn-meta (fn [old new ^MetaData meta]
+                            (if (= (.last-offset meta) old) (assoc meta :last-offset new) meta))
+        update-pt-fn-bin (fn [old new bin-indexs]
+                           (let [bin-keys (keys bin-indexs)
+                                 ch-vals (vals bin-indexs)
+                                 update-fn (partial (fn [o n chs]
+                                                      (->> chs
+                                                           (map (partial (fn [o' n' ^Chunk ch]
+                                                                           (if (= (.end ch) o')
+                                                                             (assoc ch :end n')
+                                                                             ch)) o n))
+                                                           (into []))) old new)]
+                             (zipmap bin-keys (map update-fn ch-vals))))]
+    (loop [i 0
+           index index]
+      (if (< i nrefs)
+        (if (get index i)
+          (recur (inc i) (-> index
+                             (update-in [i :meta-data] (partial update-pt-fn-meta oldpt newpt))
+                             (update-in [i :bin-index] (partial update-pt-fn-bin oldpt newpt))))
+          (recur (inc i) index))
+        index))))
```

Comment:
https://github.com/bbatsov/clojure-style-guide#higher-order-functions

Don't use `loop`-`recur` unless you need to optimize the function really heavily.
We can use higher order functions and macros instead to improve readability like the followings:

```clojure
(defn update-last-pointer [index last-buf-ptr eof-ptr]
  (reduce
   (fn [r [ref-id bin i]]
     (-> r
         (assoc-in [ref-id :bin-index bin i :end] eof-ptr)
         (assoc-in [ref-id :meta-data :last-offset] eof-ptr)))
   index
   (for [[ref-id {:keys [bin-index]}] (dissoc index :no-coordinate-alns)
         [bin chunks] bin-index
         [i {:keys [end]}] (map-indexed vector chunks)
         :when (= end last-buf-ptr)]
     [ref-id bin i])))
```

or maybe we can write it in a more direct way:

```clojure
(defn update-last-pointer [index eof-ptr]
  (if (pos? (get index :no-coordinate-alns 0))
    index
    (let [last-ref (apply max (keys (dissoc index :no-coordinate-alns)))
          last-key (->> (for [[bin chunks] (get-in index [last-ref :bin-index])
                              [i {:keys [end]}] (map-indexed vector chunks)]
                          [end [last-ref :bin-index bin i :end]])
                        (apply max-key first)
                        last)]
      (-> index
          (assoc-in [last-ref :meta-data :last-offset] eof-ptr)
          (assoc-in last-key eof-ptr)))))
```

(Please correct me if I'm wrong 🙏 )

Reply:
Fixed.